api/python/t4/packages.py

```diff
@@ -617,30 +596,29 @@ def set(self, logical_key, entry=None, meta=None):
         Returns:
             self
         """
-        if entry is None and meta is None:
-            raise PackageException('Must specify either entry or meta')
-
-        if entry is None:
-            return self._update_meta(logical_key, meta)
-
-        if isinstance(entry, (string_types, binary_type, getattr(os, 'PathLike'))):
+        if isinstance(entry, (string_types, binary_type, getattr(os, 'PathLike', str))):
             entry = PackageEntry.from_local_path(entry)
-            if meta is not None:
-                entry.meta = meta
-            self._data[logical_key] = entry
         elif isinstance(entry, PackageEntry):
-            if meta is not None:
-                raise PackageException("Must specify metadata in the entry")
-            self._data[logical_key] = entry
+            entry = entry._clone()
         else:
-            raise NotImplementedError("Needs to be of type str")
+            raise TypeError("Expected a string for entry")
 
-        return self
+        if meta is not None:
+            entry.meta = meta
+
+        path = self._split_key(logical_key)
+
+        pkg = self._ensure_subpackage(path[:-1])
+        pkg._children[path[-1]] = entry
 
-    def _update_meta(self, logical_key, meta):
-        self._data[logical_key].meta = meta
         return self
 
+    def _ensure_subpackage(self, path):
```

Comment:
Please write a docstring that specifies at least the types of the arguments and the type this returns